### PR TITLE
CP-43949: Raise exception when we can't find networkd.db

### PR DESCRIPTION
**Pull request by GeraldEV**
This covers the case where we have a management interface (mgmt_iface) from /etc/xensource-inventory but we can't find networkd.db in the installation (in the case where we've started an install/upgrade but it was interrupted)

net-admin-bridge or net-admin-interface is required by performInstallation, the failure if they are missing is abstract and unhelpful, this change will fail earlier and more clearly

### CODE REVIEW

**Review by bernhardkaindl — APPROVED** [review]

File: product.py
Comment (rosslagerwall) [inline product.py:284]: As these exceptions may be presented to the user in some situations, I would suggest keeping the text higher level - e.g. `Could not find network configuration`  - and then logging a message with the details.

Comment (rosslagerwall) [reply product.py:284]: I think you need to raise an exception in the `if not mgmt_iface:` clause as well.

**Review by rosslagerwall — COMMENTED** [review]
Have you checked whether the callers handle the exception properly?

E.g. `setup_runtime_networking()` just ignores any exception completely (though I think in that case it might be OK because if the information can't be retrieved it will ask the user for it interactively).

File: product.py
Comment (GeraldEV) [reply product.py]: I think I've appropriately caught and handled the exception everywhere (it gets quite nested quickly)
Some places are for checking if we need to update the networking which is skipped if we don't have the information to update in the settings
Updated the exceptions to be user friendly and added logging

**Review by rosslagerwall — DISMISSED** [review]
Can you squash fixup commits before merging? Thanks

File: product.py
Comment (bernhardkaindl) [inline product.py:285]: I see a typo here, what about:
```suggestion
                logger.log("Failed to find " + self.join_state_path(constants.NETWORK_DB))
```

File: tui/installer/__init__.py
Comment (bernhardkaindl) [inline tui/installer/__init__.py:86]: ditto
```suggestion
        if (
            "installation-to-overwrite" in answers
            and answers["installation-to-overwrite"].settingsAvailable()
        ):
```

File: tui/installer/__init__.py
Comment (bernhardkaindl) [inline tui/installer/__init__.py:91]: ditto / likewise, and
the `key not in dict` pattern and together with `or not` is also easier or less difficult to read for me:
```suggestion
        if (
            "installation-to-overwrite" not in answers
            or not answers["installation-to-overwrite"].settingsAvailable()
        ):
```

**Review by bernhardkaindl — CHANGES_REQUESTED** [review]
@GeraldEV Would these suggestions good for you?

If these are not OK, please notify me.

Of course, I could revert the status to Approved otherwise.

The typo in the log message would be good to fix anyway.

File: tui/installer/__init__.py
Comment (GeraldEV) [reply tui/installer/__init__.py]: I've opted for the `and \` convention in keeping with the rest of this code base
Same for other comments

File: tui/installer/__init__.py
Comment (bernhardkaindl) [inline tui/installer/__init__.py:94]: When the `not` is in the individual conditions (language-like), the `and` here needs to be an `or`.

With the `and`, the 2nd condition would be attempted when `'installation-to-overwrite' not in answers` ;-)

I'd also like to mention or discuss formatting in general:

I while ago, I was actively contributing as one of the maintainers of `https://spack.io/` (now I'm just a maintainer on leave, one could say), I was initially not so happy with the team's decision to use the black formatter for all code (and it's a huge Python code base, the complete code was re-formatted using black).

I learnt that while Black being an opinionated code formatter (having no options besides the line length and if string quoting shall be normalized to `"`-strings), it's formatting is actually very readable.

The good thing about black's formatting is, that there the formatting style is clear, there is no ambiguity how black formats it, it's always consistent.

Which is also good in another way: When using black, black formats the code, and the only variable is the maximum line length to use for black formatting.

This allows to not even have to thinking or weighting on how the code should be formatted, as there is one way it will look like after black formatted it.

Now, Of course, we can't reformat all code at once using black, as most larger patch sets would have conflicts at that point, but I'd be in favour of comitting new changes using the formatting from the tool `darker`, which formats only the added lines using black, and using it since many months, I found not even a single problem doing so.

I know, hat the formatting which black/darker provide for multi-line conditions looks as first foreign. But after some time, I've got used to it:

At least to me, it makes the such multi-line if condition more clearer to read, as the
```py
    if (
```
and the closing 
```py
    ):
```
give room to better see what's in the if condition, and the indentation of the two lines is not indented by 8 characters, they are on the same indentation level, which is good:
```suggestion
        if (
            "installation-to-overwrite" not in answers
            or not answers["installation-to-overwrite"].settingsAvailable()
        ):
```
Even if that causes two more lines of code, I think they are worth it to catch such errors better.

**Review by bernhardkaindl — CHANGES_REQUESTED** [review]

**Review by bernhardkaindl — APPROVED** [review]

**Review by rosslagerwall — DISMISSED** [review]

**Review by rosslagerwall — APPROVED** [review]

**Review by bernhardkaindl — APPROVED** [review]

**Pull request merged**
